File: go/memory/correction.go
// SPDX-License-Identifier: Apache-2.0

package memory

import (
	"fmt"
	"regexp"
	"strings"
)

// Correction is the structured result of DetectCorrection.
type Correction struct {
	Snippet string
	Phrase  string
}

// CorrectionReminderOptions configures BuildCorrectionReminderWithOptions.
type CorrectionReminderOptions struct {
	SearchTool    string
	CreateTool    string
	UpdateTool    string
	RemoveTool    string
	MentionChange bool
}

var correctionPhrases = []string{
	"that's wrong",
	"thats wrong",
	"you got that wrong",
	"you have that wrong",
	"you're wrong",
	"youre wrong",
	"you are wrong",
	"actually it's",
	"actually its",
	"actually it is",
	"it's actually",
	"its actually",
	"it is actually",
	"stop saying",
	"don't say",
	"do not say",
	"not right",
	"wrong, it's",
	"wrong, its",
	"wrong it's",
	"i never said",
	"i didn't say",
	"i did not say",
	"that's not",
	"thats not",
	"that is not",
	"correct that",
	"correction:",
	"please correct",
	"please update",
	"please remove",
	"please forget",
	"forget that",
	"forget what",
	"remember instead",
	"the correct",
}

var soloCorrectionPatterns = []*regexp.Regexp{
	regexp.MustCompile(`(?i)^\s*no[\s,\-]+(it'?s|the|that|it is|its|that is|its actually)\b`),
	regexp.MustCompile(`(?i)^\s*(wrong|incorrect)[\s,!\.]`),
	regexp.MustCompile(`(?i)^\s*actually[,\s\-]`),
}

var falsePositiveSubstrings = []string{
	"no problem",
	"no idea",
	"no worries",
	"no rush",
	"no big deal",
	"wrong end of the stick",
	"wrong number",
	"wrong place",
	"nothing wrong",
	"not wrong",
	"no, before that",
	"no, but",
}

// DetectCorrection returns a Correction when latestUserText looks like a
// correction of a previously stated fact or preference.
func DetectCorrection(latestUserText string) (Correction, bool) {
	if latestUserText == "" {
		return Correction{}, false
	}

	clean := strings.ToLower(latestUserText)
	clean = strings.TrimSpace(clean)
	if clean == "" {
		return Correction{}, false
	}

	for _, fp := range falsePositiveSubstrings {
		if strings.Contains(clean, fp) {
			return Correction{}, false
		}
	}

	for _, phrase := range correctionPhrases {
		if idx := strings.Index(clean, phrase); idx >= 0 {
			return Correction{
				Snippet: extractCorrectionSnippet(latestUserText, idx),
				Phrase:  phrase,
			}, true
		}
	}

	for _, re := range soloCorrectionPatterns {
		if loc := re.FindStringIndex(clean); loc != nil {
			return Correction{
				Snippet: extractCorrectionSnippet(latestUserText, loc[0]),
				Phrase:  re.String(),
			}, true
		}
	}

	return Correction{}, false
}

// BuildCorrectionReminder returns a default memory-tool reminder for a
// correction-looking turn, or the empty string when no correction is detected.
func BuildCorrectionReminder(userInput string) string {
	return BuildCorrectionReminderWithOptions(userInput, CorrectionReminderOptions{
		SearchTool:    "memory_search",
		UpdateTool:    "memory_update",
		RemoveTool:    "memory_remove",
		CreateTool:    "memory_create",
		MentionChange: true,
	})
}

// BuildCorrectionReminderWithOptions returns a per-turn instruction that
// nudges an assistant to repair stale memory before answering.
func BuildCorrectionReminderWithOptions(userInput string, opts CorrectionReminderOptions) string {
	c, ok := DetectCorrection(userInput)
	if !ok {
		return ""
	}
	snippet := c.Snippet
	if snippet == "" {
		snippet = userInput
	}

	searchTool := defaultString(opts.SearchTool, "memory_search")
	updateTool := defaultString(opts.UpdateTool, "memory_update")
	removeTool := defaultString(opts.RemoveTool, "memory_remove")
	createTool := defaultString(opts.CreateTool, "memory_create")

	reminder := fmt.Sprintf(
		"User correction detected (%q). Before answering, call %s for the disputed topic. If a stale entry exists, call %s or %s with a reason. If no entry exists yet but the correction is durable, call %s.",
		snippet,
		searchTool,
		updateTool,
		removeTool,
		createTool,
	)
	if opts.MentionChange {
		reminder += " Mention in your reply what you changed."
	}
	return reminder
}

func extractCorrectionSnippet(original string, matchStart int) string {
	const maxLen = 160
	runes := []rune(original)
	if len(runes) <= maxLen {
		return strings.TrimSpace(string(runes))
	}

	runeStart := 0
	for i := range original {
		if i >= matchStart {
			break
		}
		runeStart++
	}
	start := runeStart - maxLen/3
	if start < 0 {
		start = 0
	}
	end := start + maxLen
	if end > len(runes) {
		end = len(runes)
		start = end - maxLen
		if start < 0 {
			start = 0
		}
	}
	prefix := ""
	if start > 0 {
		prefix = "..."
	}
	suffix := ""
	if end < len(runes) {
		suffix = "..."
	}
	return strings.TrimSpace(prefix + string(runes[start:end]) + suffix)
}

func defaultString(value, fallback string) string {
	value = strings.TrimSpace(value)
	if value == "" {
		return fallback
	}
	return value
}

File: sdks/ts/memory/src/memory/correction.ts
// SPDX-License-Identifier: Apache-2.0

export type Correction = {
  readonly snippet: string
  readonly phrase: string
}

export type CorrectionReminderOptions = {
  readonly searchTool?: string
  readonly createTool?: string
  readonly updateTool?: string
  readonly removeTool?: string
  readonly mentionChange?: boolean
}

const CORRECTION_PHRASES = [
  "that's wrong",
  'thats wrong',
  'you got that wrong',
  'you have that wrong',
  "you're wrong",
  'youre wrong',
  'you are wrong',
  "actually it's",
  'actually its',
  'actually it is',
  "it's actually",
  'its actually',
  'it is actually',
  'stop saying',
  "don't say",
  'do not say',
  'not right',
  "wrong, it's",
  'wrong, its',
  "wrong it's",
  'i never said',
  "i didn't say",
  'i did not say',
  "that's not",
  'thats not',
  'that is not',
  'correct that',
  'correction:',
  'please correct',
  'please update',
  'please remove',
  'please forget',
  'forget that',
  'forget what',
  'remember instead',
  'the correct',
] as const

const SOLO_CORRECTION_PATTERNS = [
  /^\s*no[\s,\-]+(it'?s|the|that|it is|its|that is|its actually)\b/i,
  /^\s*(wrong|incorrect)[\s,!.]/i,
  /^\s*actually[,\s\-]/i,
] as const

const FALSE_POSITIVE_SUBSTRINGS = [
  'no problem',
  'no idea',
  'no worries',
  'no rush',
  'no big deal',
  'wrong end of the stick',
  'wrong number',
  'wrong place',
  'nothing wrong',
  'not wrong',
  'no, before that',
  'no, but',
] as const

export const detectCorrection = (latestUserText: string): Correction | undefined => {
  if (latestUserText === '') return undefined

  const clean = latestUserText.toLowerCase().trim()
  if (clean === '') return undefined

  for (const falsePositive of FALSE_POSITIVE_SUBSTRINGS) {
    if (clean.includes(falsePositive)) return undefined
  }

  for (const phrase of CORRECTION_PHRASES) {
    const index = clean.indexOf(phrase)
    if (index >= 0) {
      return {
        snippet: extractCorrectionSnippet(latestUserText, index),
        phrase,
      }
    }
  }

  for (const pattern of SOLO_CORRECTION_PATTERNS) {
    const match = pattern.exec(clean)
    if (match?.index !== undefined) {
      return {
        snippet: extractCorrectionSnippet(latestUserText, match.index),
        phrase: pattern.source,
      }
    }
  }

  return undefined
}

export const buildCorrectionReminder = (userInput: string): string =>
  buildCorrectionReminderWithOptions(userInput, {
    searchTool: 'memory_search',
    updateTool: 'memory_update',
    removeTool: 'memory_remove',
    createTool: 'memory_create',
    mentionChange: true,
  })

export const buildCorrectionReminderWithOptions = (
  userInput: string,
  opts: CorrectionReminderOptions = {},
): string => {
  const correction = detectCorrection(userInput)
  if (correction === undefined) return ''

  const snippet = correction.snippet || userInput
  const searchTool = defaultString(opts.searchTool, 'memory_search')
  const updateTool = defaultString(opts.updateTool, 'memory_update')
  const removeTool = defaultString(opts.removeTool, 'memory_remove')
  const createTool = defaultString(opts.createTool, 'memory_create')

  let reminder = `User correction detected (${JSON.stringify(snippet)}). Before answering, call ${searchTool} for the disputed topic. If a stale entry exists, call ${updateTool} or ${removeTool} with a reason. If no entry exists yet but the correction is durable, call ${createTool}.`
  if (opts.mentionChange === true) {
    reminder += ' Mention in your reply what you changed.'
  }
  return reminder
}

const extractCorrectionSnippet = (original: string, matchStart: number): string => {
  const maxLen = 160
  const runes = Array.from(original)
  if (runes.length <= maxLen) return original.trim()

  const runeStart = Array.from(original.slice(0, matchStart)).length
  let start = runeStart - Math.floor(maxLen / 3)
  if (start < 0) start = 0

  let end = start + maxLen
  if (end > runes.length) {
    end = runes.length
    start = Math.max(0, end - maxLen)
  }

  const prefix = start > 0 ? '...' : ''
  const suffix = end < runes.length ? '...' : ''
  return `${prefix}${runes.slice(start, end).join('')}${suffix}`.trim()
}

const defaultString = (value: string | undefined, fallback: string): string => {
  const trimmed = value?.trim()
  return trimmed && trimmed !== '' ? trimmed : fallback
}
